src/main/views/MattermostView.ts

```diff
@@ -183,28 +177,65 @@ export class MattermostView extends EventEmitter {
         });
     }
 
-    get serverUrl() {
-        let url = `${this.tab.server.url}`;
-        if (url.endsWith('/')) {
-            url = url.slice(0, url.length - 1);
+    private extractCORSHeaders = (details: OnBeforeSendHeadersListenerDetails) => {
+        if (!details.url.startsWith(this.tab.server.url.origin)) {
+            return {} as Headers;
         }
-        return url;
-    }
 
-    private addNoCacheForRemoteEntryRequest = (details: OnBeforeSendHeadersListenerDetails) => {
-        log.silly('WindowManager.addNoCacheForRemoteEntry', details.requestHeaders);
+        if (details.method !== 'OPTIONS') {
+            return {} as Headers;
+        }
 
-        if (!details.url.match(new RegExp(`${this.serverUrl}/static/remote_entry.js`))) {
+        if (details.requestHeaders.Origin !== `mm-desktop://${this.tab.server.url.host}`) {
             return {} as Headers;
         }
 
+        const headers = [...details.requestHeaders['Access-Control-Request-Headers'].split(',')];
+        headers.forEach((header) => {
+            if (!this.corsHeaders.includes(header)) {
+                this.corsHeaders.push(header);
+            }
+        });
+        if (!this.corsMethods.includes(details.requestHeaders['Access-Control-Request-Method'])) {
+            this.corsMethods.push(details.requestHeaders['Access-Control-Request-Method']);
+        }
+
+        return {};
+    }
+
+    private addCORSResponseHeader = (details: OnHeadersReceivedListenerDetails): HeadersReceivedResponse => {
+        if (!details.url.startsWith(this.tab.server.url.origin)) {
```

Comment:
This check is broken and can be bypassed with XHR and credentials in URLs. If a user adds two servers, `mattermost.evil.com` and `community.mattermost.com`, the former can then do this:

```js
let xhr = new XMLHttpRequest();
xhr.open("GET", "https://mattermost.evil.com@community.mattermost.com/any/path");
xhr.send(null);
```

CORS headers will get added and the response will come through to the attacker. This needs to be fixed before merging.

Reply:
I'll replace this with an exact origin check, that should fix this bug.